File: diffmerge-3/py/diffmerge/patch.py
"""Patch creation and application for the diffmerge library."""

import re
from typing import Optional, Dict, Any, List, Tuple
from .types import (
    PatchHunk, PatchLine, ParsedPatch, ApplyResult, PatchOptions, ParseError
)
from .diff import _compute_lcs, _split_lines


def _create_unified_hunks(old_lines: List[str], new_lines: List[str],
                          context_lines: int = 3) -> List[PatchHunk]:
    """Create unified diff hunks from old and new lines."""
    lcs = _compute_lcs(old_lines, new_lines)

    # Build edit script: list of (op, old_idx, new_idx, content)
    edits = []
    old_idx = 0
    new_idx = 0
    lcs_idx = 0

    while old_idx < len(old_lines) or new_idx < len(new_lines):
        if lcs_idx < len(lcs):
            lcs_old, lcs_new = lcs[lcs_idx]

            # Deletions before LCS match
            while old_idx < lcs_old:
                edits.append(("-", old_idx, None, old_lines[old_idx]))
                old_idx += 1

            # Insertions before LCS match
            while new_idx < lcs_new:
                edits.append(("+", None, new_idx, new_lines[new_idx]))
                new_idx += 1

            # Equal line (LCS match)
            edits.append((" ", old_idx, new_idx, old_lines[old_idx]))
            old_idx += 1
            new_idx += 1
            lcs_idx += 1
        else:
            # Remaining deletions
            while old_idx < len(old_lines):
                edits.append(("-", old_idx, None, old_lines[old_idx]))
                old_idx += 1

            # Remaining insertions
            while new_idx < len(new_lines):
                edits.append(("+", None, new_idx, new_lines[new_idx]))
                new_idx += 1

    if not edits:
        return []

    # Find change regions and group into hunks with context
    hunks = []
    change_indices = [i for i, e in enumerate(edits) if e[0] != " "]

    if not change_indices:
        return []

    # Group changes into hunks
    hunk_groups = []
    current_group = [change_indices[0]]

    for i in range(1, len(change_indices)):
        prev_idx = change_indices[i - 1]
        curr_idx = change_indices[i]

        # If changes are close enough, include in same hunk
        if curr_idx - prev_idx <= 2 * context_lines + 1:
            current_group.append(curr_idx)
        else:
            hunk_groups.append(current_group)
            current_group = [curr_idx]

    hunk_groups.append(current_group)

    # Build hunks from groups
    for group in hunk_groups:
        first_change = group[0]
        last_change = group[-1]

        # Add context before and after
        start = max(0, first_change - context_lines)
        end = min(len(edits), last_change + context_lines + 1)

        # Calculate old/new line ranges
        old_start = None
        new_start = None
        old_count = 0
        new_count = 0
        lines = []

        for i in range(start, end):
            op, oi, ni, content = edits[i]

            if op == " ":
                if old_start is None:
                    old_start = oi + 1
                if new_start is None:
                    new_start = ni + 1
                old_count += 1
                new_count += 1
            elif op == "-":
                if old_start is None:
                    old_start = oi + 1
                old_count += 1
            elif op == "+":
                if new_start is None:
                    new_start = ni + 1
                new_count += 1

            lines.append(PatchLine(op=op, content=content))

        # Handle edge case where start is 0
        if old_start is None:
            old_start = 1
        if new_start is None:
            new_start = 1

        hunks.append(PatchHunk(
            old_start=old_start,
            old_count=old_count,
            new_start=new_start,
            new_count=new_count,
            lines=lines
        ))

    return hunks


def create_patch(old: str, new: str, options: Optional[Dict[str, Any]] = None) -> str:
    """Generate unified diff format patch.

    Args:
        old: Original content
        new: New content
        options: Optional dict with:
            - old_file: Name of old file
            - new_file: Name of new file
            - context_lines: Number of context lines

    Returns:
        Unified diff format string
    """
    opts = PatchOptions()
    if options:
        if "old_file" in options:
            opts.old_file = options["old_file"]
        if "new_file" in options:
            opts.new_file = options["new_file"]
        if "context_lines" in options:
            opts.context_lines = options["context_lines"]

    old_lines = _split_lines(old)
    new_lines = _split_lines(new)

    hunks = _create_unified_hunks(old_lines, new_lines, opts.context_lines)

    # Build unified diff output
    lines = []
    lines.append(f"--- {opts.old_file}")
    lines.append(f"+++ {opts.new_file}")

    for hunk in hunks:
        # Hunk header
        header = f"@@ -{hunk.old_start},{hunk.old_count} +{hunk.new_start},{hunk.new_count} @@"
        lines.append(header)

        # Hunk lines
        for pl in hunk.lines:
            lines.append(f"{pl.op}{pl.content}")

    return "\n".join(lines) + ("\n" if lines else "")


def parse_patch(patch: str) -> dict:
    """Parse unified diff format into structured data.

    Args:
        patch: Unified diff format string

    Returns:
        Dict with old_file, new_file, and hunks

    Raises:
        ParseError: If patch format is invalid
    """
    lines = patch.split("\n")
    if not lines:
        raise ParseError("Empty patch")

    old_file = ""
    new_file = ""
    hunks = []
    current_hunk = None
    i = 0

    while i < len(lines):
        line = lines[i]

        if line.startswith("---"):
            # Old file header
            old_file = line[4:].strip()
        elif line.startswith("+++"):
            # New file header
            new_file = line[4:].strip()
        elif line.startswith("@@"):
            # Hunk header
            match = re.match(r"@@ -(\d+)(?:,(\d+))? \+(\d+)(?:,(\d+))? @@", line)
            if not match:
                raise ParseError(f"Invalid hunk header: {line}")

            old_start = int(match.group(1))
            old_count = int(match.group(2)) if match.group(2) else 1
            new_start = int(match.group(3))
            new_count = int(match.group(4)) if match.group(4) else 1

            if current_hunk:
                hunks.append(current_hunk)

            current_hunk = PatchHunk(
                old_start=old_start,
                old_count=old_count,
                new_start=new_start,
                new_count=new_count,
                lines=[]
            )
        elif current_hunk is not None:
            if line.startswith(" "):
                current_hunk.lines.append(PatchLine(op=" ", content=line[1:]))
            elif line.startswith("+"):
                current_hunk.lines.append(PatchLine(op="+", content=line[1:]))
            elif line.startswith("-"):
                current_hunk.lines.append(PatchLine(op="-", content=line[1:]))
            elif line == "":
                # Empty line at end of patch
                pass
            else:
                # Treat as context line if no prefix and hunk is active
                pass

        i += 1

    if current_hunk:
        hunks.append(current_hunk)

    if not old_file and not new_file and not hunks:
        raise ParseError("Invalid patch format")

    result = ParsedPatch(old_file=old_file, new_file=new_file, hunks=hunks)
    return result.to_dict()


def _apply_hunk(lines: List[str], hunk: PatchHunk, offset: int) -> Tuple[bool, List[str], int]:
    """Apply a single hunk to lines.

    Returns (success, new_lines, new_offset)
    """
    # Calculate the position to apply (1-indexed to 0-indexed, plus offset)
    pos = hunk.old_start - 1 + offset

    # Verify context lines match
    expected_old = []
    for pl in hunk.lines:
        if pl.op in (" ", "-"):
            expected_old.append(pl.content)

    # Check if hunk applies at this position
    can_apply = True
    if pos < 0:
        can_apply = False
    elif pos + len(expected_old) > len(lines):
        # Try fuzzy matching
        can_apply = False

    if can_apply:
        # Verify lines match
        for i, expected in enumerate(expected_old):
            if pos + i >= len(lines):
                can_apply = False
                break
            if lines[pos + i] != expected:
                can_apply = False
                break

    if not can_apply:
        # Try fuzzy matching within a range
        for fuzz_offset in range(-3, 4):
            fuzz_pos = pos + fuzz_offset
            if fuzz_pos < 0 or fuzz_pos + len(expected_old) > len(lines):
                continue

            match = True
            for i, expected in enumerate(expected_old):
                if lines[fuzz_pos + i] != expected:
                    match = False
                    break

            if match:
                pos = fuzz_pos
                can_apply = True
                offset += fuzz_offset
                break

    if not can_apply:
        return False, lines, offset

    # Apply the hunk
    new_lines = lines[:pos]
    for pl in hunk.lines:
        if pl.op == " ":
            new_lines.append(pl.content)
        elif pl.op == "+":
            new_lines.append(pl.content)
        # Skip "-" lines (deletions)

    new_lines.extend(lines[pos + len(expected_old):])

    # Adjust offset for line count changes
    new_count = sum(1 for pl in hunk.lines if pl.op in (" ", "+"))
    offset += new_count - len(expected_old)

    return True, new_lines, offset


def apply_patch(content: str, patch: str) -> dict:
    """Apply a unified diff patch to content.

    Args:
        content: Original content
        patch: Unified diff patch

    Returns:
        Dict with content, success, hunks_applied, hunks_failed, errors
    """
    try:
        parsed = parse_patch(patch)
    except ParseError as e:
        return ApplyResult(
            content=content,
            success=False,
            hunks_applied=0,
            hunks_failed=1,
            errors=[str(e)]
        ).to_dict()

    if not parsed["hunks"]:
        return ApplyResult(
            content=content,
            success=True,
            hunks_applied=0,
            hunks_failed=0,
            errors=[]
        ).to_dict()

    lines = _split_lines(content)
    offset = 0
    hunks_applied = 0
    hunks_failed = 0
    errors = []

    for hunk_dict in parsed["hunks"]:
        hunk = PatchHunk(
            old_start=hunk_dict["old_start"],
            old_count=hunk_dict["old_count"],
            new_start=hunk_dict["new_start"],
            new_count=hunk_dict["new_count"],
            lines=[PatchLine(op=l["op"], content=l["content"]) for l in hunk_dict["lines"]]
        )

        success, lines, offset = _apply_hunk(lines, hunk, offset)
        if success:
            hunks_applied += 1
        else:
            hunks_failed += 1
            errors.append(f"Failed to apply hunk at line {hunk.old_start}")

    # Reconstruct content
    result_content = "\n".join(lines)
    if lines and content.endswith("\n"):
        result_content += "\n"

    return ApplyResult(
        content=result_content,
        success=hunks_failed == 0,
        hunks_applied=hunks_applied,
        hunks_failed=hunks_failed,
        errors=errors
    ).to_dict()


def reverse_patch(patch: str) -> str:
    """Reverse a patch (swap additions and deletions).

    Args:
        patch: Unified diff patch

    Returns:
        Reversed patch string
    """
    lines = patch.split("\n")
    result = []

    for line in lines:
        if line.startswith("---"):
            # Swap to +++
            result.append("+++" + line[3:])
        elif line.startswith("+++"):
            # Swap to ---
            result.append("---" + line[3:])
        elif line.startswith("@@"):
            # Swap old and new in hunk header
            match = re.match(r"@@ -(\d+)(?:,(\d+))? \+(\d+)(?:,(\d+))? @@(.*)", line)
            if match:
                old_start = match.group(1)
                old_count = match.group(2) if match.group(2) else "1"
                new_start = match.group(3)
                new_count = match.group(4) if match.group(4) else "1"
                rest = match.group(5) or ""
                result.append(f"@@ -{new_start},{new_count} +{old_start},{old_count} @@{rest}")
            else:
                result.append(line)
        elif line.startswith("+"):
            # Swap + to -
            result.append("-" + line[1:])
        elif line.startswith("-"):
            # Swap - to +
            result.append("+" + line[1:])
        else:
            result.append(line)

    # Fix the order: put --- before +++
    final = []
    i = 0
    while i < len(result):
        if result[i].startswith("+++") and i + 1 < len(result) and result[i + 1].startswith("---"):
            final.append(result[i + 1])
            final.append(result[i])
            i += 2
        else:
            final.append(result[i])
            i += 1

    return "\n".join(final)

File: diffmerge-3/go/dm/patch.go
package dm

import (
	"fmt"
	"regexp"
	"strconv"
	"strings"
)

// CreatePatch generates a unified diff format patch
func CreatePatch(old, new string, options PatchOptions) string {
	oldFile := options.OldFile
	newFile := options.NewFile
	contextLines := options.ContextLines

	if oldFile == "" {
		oldFile = "a"
	}
	if newFile == "" {
		newFile = "b"
	}
	if contextLines == 0 {
		contextLines = 3
	}

	oldLines := SplitLines(old)
	newLines := SplitLines(new)

	// Compute the diff
	diffResult := DiffLines(old, new, DiffOptions{})

	// If no changes, return minimal patch
	if diffResult.Stats.Additions == 0 && diffResult.Stats.Deletions == 0 {
		return fmt.Sprintf("--- %s\n+++ %s\n", oldFile, newFile)
	}

	// Build patch output
	var sb strings.Builder
	sb.WriteString(fmt.Sprintf("--- %s\n", oldFile))
	sb.WriteString(fmt.Sprintf("+++ %s\n", newFile))

	// Group hunks by regions of changes
	patchHunks := buildPatchHunks(oldLines, newLines, diffResult.Hunks, contextLines)

	for _, hunk := range patchHunks {
		sb.WriteString(hunk)
	}

	return sb.String()
}

// buildPatchHunks creates patch hunk strings from diff hunks
func buildPatchHunks(oldLines, newLines []string, hunks []DiffHunk, contextLines int) []string {
	if len(hunks) == 0 {
		return []string{}
	}

	// Find ranges of changes and group with context
	type changeRange struct {
		oldStart, oldEnd int
		newStart, newEnd int
	}

	// Collect change ranges
	ranges := []changeRange{}
	for _, h := range hunks {
		if h.Op == DiffOpDelete || h.Op == DiffOpInsert {
			oldIdx := 0
			newIdx := 0
			if h.Op == DiffOpDelete {
				oldIdx = h.OldStart - 1
			}
			if h.Op == DiffOpInsert {
				newIdx = h.NewStart - 1
			}

			if len(ranges) == 0 {
				ranges = append(ranges, changeRange{
					oldStart: oldIdx, oldEnd: oldIdx,
					newStart: newIdx, newEnd: newIdx,
				})
			} else {
				last := &ranges[len(ranges)-1]
				if h.Op == DiffOpDelete {
					if oldIdx <= last.oldEnd+contextLines*2+1 {
						last.oldEnd = oldIdx
					} else {
						ranges = append(ranges, changeRange{
							oldStart: oldIdx, oldEnd: oldIdx,
							newStart: newIdx, newEnd: newIdx,
						})
					}
				}
				if h.Op == DiffOpInsert {
					// Check proximity using the previous context
					if len(ranges) > 0 && newIdx <= ranges[len(ranges)-1].newEnd+contextLines*2+1 {
						ranges[len(ranges)-1].newEnd = newIdx
					} else if len(ranges) == 0 || ranges[len(ranges)-1].oldEnd != ranges[len(ranges)-1].oldStart ||
						ranges[len(ranges)-1].newEnd != ranges[len(ranges)-1].newStart {
						ranges = append(ranges, changeRange{
							oldStart: oldIdx, oldEnd: oldIdx,
							newStart: newIdx, newEnd: newIdx,
						})
					}
				}
			}
		}
	}

	// Simpler approach: iterate through hunks and build patch lines
	result := []string{}

	// Track old and new line positions
	oldPos := 0
	newPos := 0
	hunkIdx := 0

	for hunkIdx < len(hunks) {
		// Find the start of the next change region
		for hunkIdx < len(hunks) && hunks[hunkIdx].Op == DiffOpEqual {
			oldPos++
			newPos++
			hunkIdx++
		}

		if hunkIdx >= len(hunks) {
			break
		}

		// Found a change, collect it and surrounding context
		changeStartIdx := hunkIdx

		// Find where this change region ends
		changeEndIdx := hunkIdx
		for changeEndIdx < len(hunks) {
			if hunks[changeEndIdx].Op == DiffOpEqual {
				// Count consecutive equals
				equalCount := 0
				for i := changeEndIdx; i < len(hunks) && hunks[i].Op == DiffOpEqual; i++ {
					equalCount++
				}
				if equalCount > contextLines*2 {
					break
				}
			}
			changeEndIdx++
		}

		// Determine hunk bounds with context
		oldHunkStart := max(0, oldPos-contextLines)
		newHunkStart := max(0, newPos-contextLines)

		// Build the hunk content
		var hunkContent strings.Builder
		hunkOldLines := 0
		hunkNewLines := 0

		// Add leading context from original
		for i := oldHunkStart; i < oldPos && i < len(oldLines); i++ {
			line := stripNewline(oldLines[i])
			hunkContent.WriteString(" " + line + "\n")
			hunkOldLines++
			hunkNewLines++
		}

		// Process the change region
		tempOldPos := oldPos
		tempNewPos := newPos
		for i := changeStartIdx; i < changeEndIdx && i < len(hunks); i++ {
			h := hunks[i]
			switch h.Op {
			case DiffOpEqual:
				line := stripNewline(h.Content)
				hunkContent.WriteString(" " + line + "\n")
				hunkOldLines++
				hunkNewLines++
				tempOldPos++
				tempNewPos++
			case DiffOpDelete:
				line := stripNewline(h.Content)
				hunkContent.WriteString("-" + line + "\n")
				hunkOldLines++
				tempOldPos++
			case DiffOpInsert:
				line := stripNewline(h.Content)
				hunkContent.WriteString("+" + line + "\n")
				hunkNewLines++
				tempNewPos++
			}
		}

		// Add trailing context
		trailingStart := tempOldPos
		trailingEnd := min(tempOldPos+contextLines, len(oldLines))
		for i := trailingStart; i < trailingEnd; i++ {
			line := stripNewline(oldLines[i])
			hunkContent.WriteString(" " + line + "\n")
			hunkOldLines++
			hunkNewLines++
		}

		// Format hunk header
		hunkHeader := fmt.Sprintf("@@ -%d,%d +%d,%d @@\n",
			oldHunkStart+1, hunkOldLines,
			newHunkStart+1, hunkNewLines)

		result = append(result, hunkHeader+hunkContent.String())

		// Move positions
		oldPos = tempOldPos
		newPos = tempNewPos
		hunkIdx = changeEndIdx

		// Skip trailing context in main loop
		for i := 0; i < contextLines && hunkIdx < len(hunks) && hunks[hunkIdx].Op == DiffOpEqual; i++ {
			oldPos++
			newPos++
			hunkIdx++
		}
	}

	return result
}

// stripNewline removes trailing newline from a line
func stripNewline(s string) string {
	return strings.TrimRight(s, "\r\n")
}

// ApplyPatch applies a unified diff patch to content
func ApplyPatch(content, patch string) ApplyResult {
	parsed, err := parsePatchInternal(patch)
	if err != nil {
		return ApplyResult{
			Content:      content,
			Success:      false,
			HunksApplied: 0,
			HunksFailed:  0,
			Errors:       []string{err.Error()},
		}
	}

	if len(parsed.Hunks) == 0 {
		return ApplyResult{
			Content:      content,
			Success:      true,
			HunksApplied: 0,
			HunksFailed:  0,
			Errors:       []string{},
		}
	}

	lines := SplitLines(content)
	result := make([]string, 0, len(lines))
	lineIdx := 0
	hunksApplied := 0
	hunksFailed := 0
	errors := []string{}

	for _, hunk := range parsed.Hunks {
		targetLine := hunk.OldStart - 1

		// Add lines before this hunk
		for lineIdx < targetLine && lineIdx < len(lines) {
			result = append(result, lines[lineIdx])
			lineIdx++
		}

		// Try to apply the hunk
		success, applied := applyHunk(lines, lineIdx, hunk)
		if success {
			result = append(result, applied...)
			// Skip over the old lines that were replaced
			for _, pl := range hunk.Lines {
				if pl.Op == " " || pl.Op == "-" {
					lineIdx++
				}
			}
			hunksApplied++
		} else {
			// Hunk failed, try fuzzy matching
			fuzzyLine, fuzzyApplied := fuzzyApplyHunk(lines, hunk)
			if fuzzyLine >= 0 {
				// Add any lines we skipped to reach the fuzzy match
				for lineIdx < fuzzyLine {
					result = append(result, lines[lineIdx])
					lineIdx++
				}
				result = append(result, fuzzyApplied...)
				for _, pl := range hunk.Lines {
					if pl.Op == " " || pl.Op == "-" {
						lineIdx++
					}
				}
				hunksApplied++
			} else {
				errors = append(errors, fmt.Sprintf("Hunk at line %d failed to apply", hunk.OldStart))
				hunksFailed++
			}
		}
	}

	// Add remaining lines
	for lineIdx < len(lines) {
		result = append(result, lines[lineIdx])
		lineIdx++
	}

	return ApplyResult{
		Content:      strings.Join(result, ""),
		Success:      hunksFailed == 0,
		HunksApplied: hunksApplied,
		HunksFailed:  hunksFailed,
		Errors:       errors,
	}
}

// applyHunk tries to apply a hunk at a specific position
func applyHunk(lines []string, startLine int, hunk PatchHunk) (bool, []string) {
	result := []string{}
	lineIdx := startLine

	for _, pl := range hunk.Lines {
		switch pl.Op {
		case " ":
			// Context line must match
			if lineIdx >= len(lines) {
				return false, nil
			}
			expectedContent := pl.Content
			actualContent := stripNewline(lines[lineIdx])
			if actualContent != expectedContent {
				return false, nil
			}
			result = append(result, lines[lineIdx])
			lineIdx++
		case "-":
			// Deletion must match
			if lineIdx >= len(lines) {
				return false, nil
			}
			expectedContent := pl.Content
			actualContent := stripNewline(lines[lineIdx])
			if actualContent != expectedContent {
				return false, nil
			}
			// Don't add to result (deleted)
			lineIdx++
		case "+":
			// Addition
			result = append(result, pl.Content+"\n")
		}
	}

	return true, result
}

// fuzzyApplyHunk tries to find the hunk content elsewhere in the file
func fuzzyApplyHunk(lines []string, hunk PatchHunk) (int, []string) {
	// Get the context lines from the hunk
	contextLines := []string{}
	for _, pl := range hunk.Lines {
		if pl.Op == " " || pl.Op == "-" {
			contextLines = append(contextLines, pl.Content)
		}
	}

	if len(contextLines) == 0 {
		return -1, nil
	}

	// Search for matching position
	for offset := 0; offset < len(lines); offset++ {
		// Try both directions from original position
		for _, dir := range []int{-1, 1} {
			searchLine := hunk.OldStart - 1 + (offset * dir)
			if searchLine < 0 || searchLine >= len(lines) {
				continue
			}

			// Check if context matches at this position
			matches := true
			tempLine := searchLine
			for _, ctx := range contextLines {
				if tempLine >= len(lines) || stripNewline(lines[tempLine]) != ctx {
					matches = false
					break
				}
				tempLine++
			}

			if matches {
				success, applied := applyHunk(lines, searchLine, hunk)
				if success {
					return searchLine, applied
				}
			}
		}
	}

	return -1, nil
}

// ReversePatch reverses a patch (swaps additions and deletions)
func ReversePatch(patch string) string {
	lines := strings.Split(patch, "\n")
	result := []string{}

	for _, line := range lines {
		if strings.HasPrefix(line, "---") {
			// Swap --- to +++
			result = append(result, "+++"+line[3:])
		} else if strings.HasPrefix(line, "+++") {
			// Swap +++ to ---
			result = append(result, "---"+line[3:])
		} else if strings.HasPrefix(line, "@@") {
			// Reverse the hunk header
			result = append(result, reverseHunkHeader(line))
		} else if strings.HasPrefix(line, "-") && !strings.HasPrefix(line, "---") {
			// Change deletion to addition
			result = append(result, "+"+line[1:])
		} else if strings.HasPrefix(line, "+") && !strings.HasPrefix(line, "+++") {
			// Change addition to deletion
			result = append(result, "-"+line[1:])
		} else {
			result = append(result, line)
		}
	}

	return strings.Join(result, "\n")
}

// reverseHunkHeader reverses the line numbers in a hunk header
func reverseHunkHeader(header string) string {
	re := regexp.MustCompile(`@@ -(\d+),?(\d*) \+(\d+),?(\d*) @@`)
	matches := re.FindStringSubmatch(header)
	if matches == nil {
		return header
	}

	oldStart := matches[1]
	oldCount := matches[2]
	if oldCount == "" {
		oldCount = "1"
	}
	newStart := matches[3]
	newCount := matches[4]
	if newCount == "" {
		newCount = "1"
	}

	return fmt.Sprintf("@@ -%s,%s +%s,%s @@", newStart, newCount, oldStart, oldCount)
}

// ParsePatch parses a unified diff format into structured data
func ParsePatch(patch string) (ParsedPatch, error) {
	return parsePatchInternal(patch)
}

// parsePatchInternal parses a unified diff
func parsePatchInternal(patch string) (ParsedPatch, error) {
	lines := strings.Split(patch, "\n")
	result := ParsedPatch{
		Hunks: []PatchHunk{},
	}

	i := 0

	// Skip empty lines at start
	for i < len(lines) && strings.TrimSpace(lines[i]) == "" {
		i++
	}

	// Parse file headers
	if i < len(lines) && strings.HasPrefix(lines[i], "---") {
		result.OldFile = strings.TrimSpace(lines[i][3:])
		i++
	}
	if i < len(lines) && strings.HasPrefix(lines[i], "+++") {
		result.NewFile = strings.TrimSpace(lines[i][3:])
		i++
	}

	// Check if we found valid headers or have any hunks
	hasHeaders := result.OldFile != "" || result.NewFile != ""
	hasHunks := false
	for j := i; j < len(lines); j++ {
		if strings.HasPrefix(lines[j], "@@") {
			hasHunks = true
			break
		}
	}

	if !hasHeaders && !hasHunks {
		return result, fmt.Errorf("invalid patch format")
	}

	// Parse hunks
	for i < len(lines) {
		line := lines[i]
		if strings.HasPrefix(line, "@@") {
			hunk, nextI, err := parseHunk(lines, i)
			if err != nil {
				return result, err
			}
			result.Hunks = append(result.Hunks, hunk)
			i = nextI
		} else {
			i++
		}
	}

	return result, nil
}

// parseHunk parses a single hunk
func parseHunk(lines []string, startIdx int) (PatchHunk, int, error) {
	header := lines[startIdx]
	re := regexp.MustCompile(`@@ -(\d+),?(\d*) \+(\d+),?(\d*) @@`)
	matches := re.FindStringSubmatch(header)
	if matches == nil {
		return PatchHunk{}, startIdx + 1, fmt.Errorf("invalid hunk header: %s", header)
	}

	oldStart, _ := strconv.Atoi(matches[1])
	oldCount := 1
	if matches[2] != "" {
		oldCount, _ = strconv.Atoi(matches[2])
	}
	newStart, _ := strconv.Atoi(matches[3])
	newCount := 1
	if matches[4] != "" {
		newCount, _ = strconv.Atoi(matches[4])
	}

	hunk := PatchHunk{
		OldStart: oldStart,
		OldCount: oldCount,
		NewStart: newStart,
		NewCount: newCount,
		Lines:    []PatchLine{},
	}

	i := startIdx + 1
	for i < len(lines) {
		line := lines[i]
		if line == "" {
			i++
			continue
		}
		if strings.HasPrefix(line, "@@") || strings.HasPrefix(line, "---") || strings.HasPrefix(line, "+++") {
			break
		}

		if len(line) > 0 {
			op := string(line[0])
			content := ""
			if len(line) > 1 {
				content = line[1:]
			}
			if op == " " || op == "+" || op == "-" {
				hunk.Lines = append(hunk.Lines, PatchLine{Op: op, Content: content})
			}
		}
		i++
	}

	return hunk, i, nil
}
